crypto/key_store_passphrase.go

```diff
@@ -0,0 +1,246 @@
+	}
+	key = new(Key)
+	key.Id = keyId
+	key.PrivateKey = ToECDSA(keyBytes)
+	return key, err
+}
+
+func (ks keyStorePassphrase) StoreKey(key *Key, auth string) (err error) {
+	authArray := []byte(auth)
+	salt := getEntropyCSPRNG(32)
+	derivedKey, err := scrypt.Key(authArray, salt, scryptN, scryptr, scryptp, scryptdkLen)
+	if err != nil {
+		return err
+	}
+
+	keyBytes := FromECDSA(key.PrivateKey)
+	keyBytesHash := Sha3(keyBytes)
+	toEncrypt := PKCS7Pad(append(keyBytes, keyBytesHash...))
+
+	AES256Block, err := aes.NewCipher(derivedKey)
+	if err != nil {
+		return err
+	}
+
+	iv := getEntropyCSPRNG(aes.BlockSize) // 16
+	AES256CBCEncrypter := cipher.NewCBCEncrypter(AES256Block, iv)
+	cipherText := make([]byte, len(toEncrypt))
+	AES256CBCEncrypter.CryptBlocks(cipherText, toEncrypt)
+
+	cipherStruct := cipherJSON{
+		salt,
+		iv,
+		cipherText,
+	}
+	keyStruct := encryptedKeyJSON{
+		*key.Id,
+		cipherStruct,
+	}
+	keyJSON, err := json.Marshal(keyStruct)
+	if err != nil {
+		return err
+	}
+
+	return WriteKeyFile(key.Id.String(), ks.keysDirPath, keyJSON)
+}
+
+func (ks keyStorePassphrase) DeleteKey(keyId *uuid.UUID, auth string) (err error) {
+	// only delete if correct passphrase is given
+	_, err = DecryptKey(ks, keyId, auth)
+	if err != nil {
+		return err
+	}
+
+	keyDirPath := path.Join(ks.keysDirPath, keyId.String())
+	return os.RemoveAll(keyDirPath)
+}
+
+func DecryptKey(ks keyStorePassphrase, keyId *uuid.UUID, auth string) (keyBytes []byte, err error) {
+	fileContent, err := GetKeyFile(ks.keysDirPath, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	keyProtected := new(encryptedKeyJSON)
+	err = json.Unmarshal(fileContent, keyProtected)
+
+	salt := keyProtected.Crypto.Salt
+
+	iv := keyProtected.Crypto.IV
+
+	cipherText := keyProtected.Crypto.CipherText
+
+	authArray := []byte(auth)
+	derivedKey, err := scrypt.Key(authArray, salt, scryptN, scryptr, scryptp, scryptdkLen)
+	if err != nil {
+		return nil, err
+	}
+
+	AES256Block, err := aes.NewCipher(derivedKey)
+	if err != nil {
+		return nil, err
+	}
+
+	AES256CBCDecrypter := cipher.NewCBCDecrypter(AES256Block, iv)
+	paddedPlainText := make([]byte, len(cipherText))
+	AES256CBCDecrypter.CryptBlocks(paddedPlainText, cipherText)
+
+	plainText := PKCS7Unpad(paddedPlainText)
+	if plainText == nil {
+		err = errors.New("Decryption failed: PKCS7Unpad failed after decryption")
+		return nil, err
+	}
+
+	keyBytes = plainText[:len(plainText)-32]
+	keyBytesHash := plainText[len(plainText)-32:]
+	if !bytes.Equal(Sha3(keyBytes), keyBytesHash) {
+		err = errors.New("Decryption failed: checksum mismatch")
+		return nil, err
+	}
+	return keyBytes, err
+}
+
+func getEntropyCSPRNG(n int) []byte {
+	mainBuff := make([]byte, n)
+	_, err := io.ReadFull(crand.Reader, mainBuff)
+	if err != nil {
+		panic("key generation: reading from crypto/rand failed: " + err.Error())
+	}
+	return mainBuff
+}
+
+// From https://leanpub.com/gocrypto/read#leanpub-auto-block-cipher-modes
+func PKCS7Pad(in []byte) []byte {
```

Comment:
Replace with https://github.com/ethereum/go-ethereum/blob/develop/ethutil/bytes.go#L170 (right padding bytes). Please be aware that `append` is generally pretty slow. `RightPadBytes` uses `make` and `copy` and returns a new slice.